### PR TITLE
use zk to support show processlist.

### PR DESCRIPTION
**Pull request by tuichenchuxin**
Fixes #9568.

- save execute process to ShowProcessListHolder.
- when execute show processlist, current instance will trigger all instances to save execute process to zookeeper.
- current instance will get execute process from zookeeper. After reading, it'll delete the data from zookeeper.

image

### CODE REVIEW

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/ShowProcessListHolder.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Show process list holder.
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class ShowProcessListHolder {
```
Comment (strongduanmu) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/ShowProcessListHolder.java:32]: @tuichenchuxin Please add final for ShowProcessListHolder.

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/subscriber/ProcessRegistrySubscriber.java [hunk 2]
```diff
@@ -56,9 +62,36 @@ public ProcessRegistrySubscriber(final ClusterPersistRepository repository) {
...
+        repository.delete(ComputeNode.getProcessTriggerNodePatch());
+        Arrays.stream(InstanceType.values()).forEach(instanceType -> {
+            Collection<String> onlineComputeNodes = repository.getChildrenKeys(ComputeNode.getOnlineNodePath(instanceType));
+            onlineComputeNodes.forEach(each -> repository.persist(ComputeNode.getProcessTriggerInstanceNodePath(each, instanceType), ""));
+        });
+    }
+    
+    @SneakyThrows
+    private void waitUntilShowProcessIsReady() {
+        while (true) {
```
Comment (strongduanmu) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/subscriber/ProcessRegistrySubscriber.java:80]: @tuichenchuxin Can we remove this loop with wait/notity?

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/subscriber/ProcessRegistrySubscriber.java [hunk 3]
```diff
@@ -78,18 +112,15 @@ public void reportExecuteProcessSummary(final ExecuteProcessSummaryReportEvent e
...
-            String executionPath = ProcessNode.getExecutionPath(executionID);
-            YamlExecuteProcessContext yamlExecuteProcessContext = YamlEngine.unmarshal(repository.get(executionPath), YamlExecuteProcessContext.class);
-            ExecuteProcessUnit executeProcessUnit = event.getExecuteProcessUnit();
-            for (YamlExecuteProcessUnit unit : yamlExecuteProcessContext.getUnitStatuses()) {
-                if (unit.getUnitID().equals(executeProcessUnit.getUnitID())) {
-                    unit.setStatus(executeProcessUnit.getStatus());
-                }
+        YamlExecuteProcessContext yamlExecuteProcessContext = ShowProcessListHolder.getInstance().get(executionID);
+        ExecuteProcessUnit executeProcessUnit = event.getExecuteProcessUnit();
+        for (YamlExecuteProcessUnit unit : yamlExecuteProcessContext.getUnitStatuses()) {
```
Comment (strongduanmu) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/subscriber/ProcessRegistrySubscriber.java:120]: Please rename unit with each.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/ShowProcessListHolder.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Show process list holder.
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class ShowProcessListHolder {
```
Comment (strongduanmu) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/ShowProcessListHolder.java:32]: @tuichenchuxin Can we rename it to ShowProcessListManager, and merge ShowProcessListLockHolder logic to this class.

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/subscriber/ProcessRegistrySubscriber.java [hunk 5]
```diff
@@ -56,8 +65,53 @@ public ProcessRegistrySubscriber(final ClusterPersistRepository repository) {
...
+    
+    @SneakyThrows
+    private boolean waitUntilShowProcessIsReady(final String showProcessListId, final Collection<String> triggerPaths) {
+        try {
+            ShowProcessListSimpleLock simpleLock = new ShowProcessListSimpleLock();
+            ShowProcessListLockHolder.getInstance().getLocks().put(showProcessListId, simpleLock);
+            int retryCount = 0;
+            while (!isReady(triggerPaths)) {
+                if (!simpleLock.doAwait(2000L)) {
+                    if (++retryCount > 4) {
```
Comment (strongduanmu) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/subscriber/ProcessRegistrySubscriber.java:97]: Please remove retryCount here.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/status/compute/watcher/ComputeNodeStateChangedWatcher.java [hunk 6]
```diff
@@ -70,10 +75,31 @@
...
         }
         return Optional.empty();
     }
     
+    private Optional<GovernanceEvent> createShowProcessListTriggerEvent(final DataChangedEvent event) {
+        Matcher matcher = getShowProcessTriggerMatcher(event);
+        if (!matcher.find()) {
+            return Optional.empty();
+        }
+        if (event.getType() == Type.ADDED) {
```
Comment (zjcnb) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/status/compute/watcher/ComputeNodeStateChangedWatcher.java:89]: @tuichenchuxin Thank you for your pull request. I think `event.getType() == Type.ADDED` can change to `Type.ADDED == event.getType()`

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/status/compute/watcher/ComputeNodeStateChangedWatcher.java [hunk 7]
```diff
@@ -70,10 +75,31 @@
...
+            return Optional.of(new ShowProcessListTriggerEvent(
+                    new InstanceDefinition(InstanceType.PROXY.name().equalsIgnoreCase(matcher.group(1)) ? InstanceType.PROXY : InstanceType.JDBC, matcher.group(2)), matcher.group(3)));
+        } else if (event.getType() == Type.DELETED) {
```
Comment (zjcnb) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/status/compute/watcher/ComputeNodeStateChangedWatcher.java:92]: Same problem.

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/ShowProcessListHolder.java (hunk 4 above)
Comment (tuichenchuxin) [reply shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/ShowProcessListHolder.java]: Thank you. Done.

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/subscriber/ProcessRegistrySubscriber.java (hunk 5 above)
Comment (tuichenchuxin) [reply shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/process/subscriber/ProcessRegistrySubscriber.java]: Thank you. Done.

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/status/compute/watcher/ComputeNodeStateChangedWatcher.java (hunk 6 above)
Comment (tuichenchuxin) [reply shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/status/compute/watcher/ComputeNodeStateChangedWatcher.java]: Thank you. Done.

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/status/compute/watcher/ComputeNodeStateChangedWatcher.java (hunk 7 above)
Comment (tuichenchuxin) [reply shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/status/compute/watcher/ComputeNodeStateChangedWatcher.java]: Thank you. Done.

File: shardingsphere-infra/shardingsphere-infra-executor/src/main/java/org/apache/shardingsphere/infra/executor/sql/process/model/yaml/YamlExecuteProcessContextPackage.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.util.Collection;
+
+/**
+ * Execute process context package for YAML.
+ */
+@NoArgsConstructor
+@Getter
+@Setter
+public final class YamlExecuteProcessContextPackage {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-executor/src/main/java/org/apache/shardingsphere/infra/executor/sql/process/model/yaml/YamlExecuteProcessContextPackage.java:32]: @tuichenchuxin Can we rename YamlExecuteProcessContextPackage to BatchYamlExecuteProcessContext?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-executor/src/main/java/org/apache/shardingsphere/infra/executor/sql/process/model/yaml/YamlExecuteProcessContextPackage.java (hunk 8 above)
Comment (tuichenchuxin) [reply shardingsphere-infra/shardingsphere-infra-executor/src/main/java/org/apache/shardingsphere/infra/executor/sql/process/model/yaml/YamlExecuteProcessContextPackage.java]: Thanks. Done.

**Review by strongduanmu — APPROVED** [review]

**Pull request merged**